### PR TITLE
Add a workflow to automatically build and publish to PyPI when a new tag is pushed.

### PR DESCRIPTION
**Pull request by aholmes**
The workflow added in this PR will build and push a new package to PyPI when a tag is pushed. The tags must follow some rules:

- The tag must start with `BL_Python.`
- The tag must end with a `-v` followed by a semantic version, e.g., `-v1.0.0`
  - The version must reflect the same version in the package's `__init__.py` file. This is not enforced at the moment.
- The name of the package must follow `BL_Python.` and precede the version, e.g., `BL_Python.web-v1.0.0`
- The name of the package must be the equivalent of the directory name under the `src/` directory for whichever package the tag is for
  - The package name for the root repository must be `all`

In addition to publishing a PyPI package, the workflow that is triggered by a new tag will also create a new release on GitHub.

### CODE REVIEW

File: .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+set -eo pipefail
+tag="${GITHUB_REF#refs/*/}"
+
+# This is the package name minus `BL_Python.`.
+package_name=$(gawk 'match($0, /BL_Python\.?([^-]+)?/, m) { print s m[1]}' <<<$tag)
```
Comment (nwiltsie) [inline .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh:6]: Interesting, I've never dealt with [here strings](https://www.gnu.org/software/bash/manual/bash.html#Here-Strings) before!

File: .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# `all` is the root of the repository and includes all other packages
+if [ "$package_name" == "all" ]; then
+    package_directory=.
+else
+    package_directory="src/$package_name/"
+fi
+echo "package_name=$package_name" >> $GITHUB_OUTPUT
+echo "package_directory=$package_directory" >> $GITHUB_OUTPUT
```
Comment (nwiltsie) [inline .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh:14]: Shellcheck and [GitHub's docs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-output-parameter) recommend you double-quote `$GITHUB_OUTPUT`.

File: .github/workflows/PyPI.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+          name: python-package-distributions
+          path: dist/
+      - name: Sign the dists with Sigstore
+        uses: sigstore/gh-action-sigstore-python@v1.2.3
+        with:
+          inputs: >-
+            ./dist/*.tar.gz
+            ./dist/*.whl
+      - name: Create GitHub Release
+        # TODO change the relase notes to use the changelog
```
Comment (nwiltsie) [inline .github/workflows/PyPI.yaml:94]:
```suggestion
        # TODO change the release notes to use the changelog
```

File: .github/workflows/PyPI.yaml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    steps:
+      - name: Download all the dists
+        uses: actions/download-artifact@v3
+        with:
+          name: python-package-distributions
+          path: dist/
+      - name: Publish distribution 📦 to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1.8
+        with:
+          repository-url: https://pypi.org/legacy/
```
Comment (nwiltsie) [inline .github/workflows/PyPI.yaml:64]: So the reference workflow only had this `with:repository-url` block on the Test PyPI upload. Is this actually needed?

Copying from https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/,

```yaml
  publish-to-pypi:
    name: >-
      Publish Python 🐍 distribution 📦 to PyPI
    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
    needs:
    - build
    runs-on: ubuntu-latest
    environment:
      name: pypi
      url: https://pypi.org/p/<package-name>  # Replace <package-name> with your PyPI project name
    permissions:
      id-token: write  # IMPORTANT: mandatory for trusted publishing

    steps:
    - name: Download all the dists
      uses: actions/download-artifact@v3
      with:
        name: python-package-distributions
        path: dist/
    - name: Publish distribution 📦 to PyPI
      uses: pypa/gh-action-pypi-publish@release/v1
```
```yaml

  publish-to-testpypi:
    name: Publish Python 🐍 distribution 📦 to TestPyPI
    needs:
    - build
    runs-on: ubuntu-latest

    environment:
      name: testpypi
      url: https://test.pypi.org/p/<package-name>

    permissions:
      id-token: write  # IMPORTANT: mandatory for trusted publishing

    steps:
    - name: Download all the dists
      uses: actions/download-artifact@v3
      with:
        name: python-package-distributions
        path: dist/
    - name: Publish distribution 📦 to TestPyPI
      uses: pypa/gh-action-pypi-publish@release/v1
      with:
        repository-url: https://test.pypi.org/legacy/
```

**Review by nwiltsie — CHANGES_REQUESTED** [review]

File: .github/workflows/PyPI.yaml (hunk 4 above)
Comment (aholmes) [reply .github/workflows/PyPI.yaml:64]: You're correct; it is not needed. I prefer to leave it in place as it is explicit. Thoughts?

Comment (nwiltsie) [reply .github/workflows/PyPI.yaml:64]: That's totally fine with me - it just deviated from the template and I wanted to confirm it was intentional.

File: .github/workflows/PyPI.yaml (hunk 3 above)
Comment (aholmes) [reply .github/workflows/PyPI.yaml]: Fixed.

File: .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh (hunk 2 above)
Comment (aholmes) [reply .github/workflows/PyPI-scripts/extract-package-details-from-tag.sh]: Fixed.

**Review by nwiltsie — APPROVED** [review]

**Pull request merged**